src/main/java/duke/CommandHandler.java

```diff
@@ -0,0 +1,220 @@
+            return Integer.parseInt(parameter) - 1;
+        }
+        if (Character.isLetter(character)) {
+            character = Character.toUpperCase(character);
+            return (int)character - Constants.LETTER_OFFSET - 1;
+        }
+        return -1;
+    }
+
+    private Task getTask(
+            MessageFormat msgFormat, CommandParser cmdParser,
+            MessageWrapper msgWrapper, ArrayList<Task> addedTasks) {
+        int index = getListedIndex(cmdParser.getParameter());
+        if (index < 0 || index >= addedTasks.size()) {
+            msgWrapper.show(
+                    "Index out of range.",
+                    msgFormat.getMessageOptions());
+            return null;
+        }
+        return addedTasks.get(index);
+    }
+
+    private void handleException(Exception e, Commands flag) {
+        String unknown = "x_x OOPS!!! I'm sorry, but I don't know what that means :-(";
+        String illegal = "x_x OOPS!!! The \"_SIGN_\" parameter of the _NAME_ command is missing.";
+        String empty = "x_x OOPS!!! The parameter \"_SIGN_\" cannot be empty.";
+        String missing = "x_x OOPS!!! The description of the \"_NAME_\" command cannot be empty.";
+        if (e instanceof EmptyCommandException) {
+            result = wrapper.wrap(
+                    empty.replace("_SIGN_", flag.SIGN.trim()),
+                    format.getMessageOptions());
+        } else if (e instanceof IllegalCommandException) {
+            result = wrapper.wrap(
+                    illegal.replace("_SIGN_", flag.SIGN.trim())
+                            .replace("_NAME_", flag.NAME.trim()),
+                    format.getMessageOptions());
+        } else if (e instanceof MissingDescriptionException) {
+            result = wrapper.wrap(
+                    missing.replace("_NAME_",flag.NAME.trim()),
+                    format.getMessageOptions());
+
+        } else if (e instanceof UnknownCommandException) {
+            result = wrapper.wrap(unknown, format.getMessageOptions());
+        } else {
+            result = wrapper.wrap(
+                    new String[]{"An unknown Exception has occurred.", e.getMessage()},
+                    format.getMessageOptions());
+        }
+    }
+
+    public boolean handleCommand(String userInput) {
+        try {
+            parser.parse(userInput);
+        } catch (Exception e) {
+            handleException(e, parser.getFlag());
+            return true; //although there is exception, still should continue.
+        }
+        Commands flag = parser.getFlag();
+        if (flag == null) {
+            return true; //although no flag, still should continue.
+        }
+        String doTaskPrefix = "Nice! I've marked this task as done:";
+        String undoTaskPrefix = "Nice! I've marked this task as undone:";
+        String cleared = "Nice! I've cleared everything in the list.";
+        String bye = " Bye. Hope to see you again soon!";
+        switch (flag) {
```

Comment:
Consider moving to your constants.java class

Reply:
I did not intend for these variables to be visible from outside this class, thus I did not put them into the Constants.java class. Maybe in the next commit I'll add in the private final modifiers to make the intention clearer.